### PR TITLE
feat: Staking V3

### PR DESCRIPTION
**Pull request by gluneau**

### CODE REVIEW

**Review by bobo-k2 — CHANGES_REQUESTED** [review]

File: src/processor.ts [hunk 1]
```diff
@@ -1,32 +1,60 @@
...
+            events.dappStaking.dAppReward.name,
         ]
     })
     .setFields({
         block: {
             timestamp: true
         }
     })
-    .setBlockRange({from: 4_342_300})
+    .setBlockRange({from: 1}) // .setBlockRange({from: 4_342_300})
```
Comment (gluneau) [inline src/processor.ts:53]: Revert before merge

File: typegen.json [hunk 2]
```diff
@@ -1,9 +1,35 @@
 {
   "outDir": "src/types",
-  "specVersions": "https://v2.archive.subsquid.io/metadata/astar",
+  "specVersions": "localNode.jsonl",
+  "specVersionsProd": "https://v2.archive.subsquid.io/metadata/astar",
```
Comment (gluneau) [inline typegen.json:4]: Revert before merge

File: src/main.ts [hunk 3]
```diff
@@ -1,203 +1,341 @@
...
+        case events.dappStaking.stake.name:
+        case events.dappStaking.unstake.name:
+        case events.dappStaking.unstakeFromUnregistered.name:
+          const stake = getStake(event);
+          entities.StakesToInsert.push(stake);
+          const dapp = await handleStakersCount(ctx, stake);
+          dapp && entities.DappsToUpdate.push(dapp);
+          break;
+
+        default:
```
[the remark this answers is not in the document]
Comment (gluneau) [reply src/main.ts:286]: Thanks, it's now handled.

**Review by bobo-k2 — CHANGES_REQUESTED** [review]

File: src/main.ts [hunk 4]
```diff
@@ -1,203 +1,354 @@
...
+  await ctx.store.upsert(entities.TvlToUpdate);
+  await ctx.store.insert(entities.StakersCountToInsert);
+  await ctx.store.upsert(entities.StakersCountToUpdate);
+  await ctx.store.insert(entities.StakesToInsert);
+  await ctx.store.upsert(entities.StakesToUpdate);
+});
+
+async function handleEvents(ctx: ProcessorContext<Store>, entities: Entities) {
+  for (let block of ctx.blocks) {
+    assert(
```
[the remark this answers is not in the document]
Comment (gluneau) [reply src/main.ts:80]: blocks will always have a timestamp, it's the data type that can be undefined

**Pull request merged**